cslib_menu improvements

CareyJWilliams opened the pull request.

CareyJWilliams committed:
Refactor cslib_menu

This should be a largely non-functional commit,
albeit a large one. The intent is to make the
module's logic a bit easier to understand by
splittings a lot of the functionality
out into internal named routines. I've also
renamed some variables and routines, and
improved the common elements.

Choice building was also restructured so that
array and simple menus can share more page logic.

CareyJWilliams committed:
Make cslib_menu testable

Adds tests for cslib_menu and the override system,
similar to what we did with *bug, to support them.

This system allows for automated choice navigation
and analysis, and will greatly increase our
confidence in the cslib_menu implementation.

diegoboston commented on mygame.js:
```diff
@@ -25,4 +25,111 @@ Scene.prototype.bug = function scene_bug(message) {
+    // Extract the next choice selection
+    var selections = this.stats.cslib_choice_select.split(",");
+    var thisSelection = parseInt(selections.shift());
+    this.stats.cslib_choice_select = selections.join(",");
```
interesting design - I guess it's the best way to interact with CS, even if a little clunky

CareyJWilliams replied:
Feel free to suggest something less clunky!

diegoboston replied:
I don't know if you can parse cslib_choice_select into a (JS) array and use that directly. Would that work as you pass it around? Or instead of splitting and rejoining, you may be able to do substring. Just ideas, not blockers.

CareyJWilliams replied:
It would work, but it would produce more lines/verbosity, I think, as you'd need to check other that and cslib_choice_select weren't empty. I think I'll leave it as-is for now, but I'm happy to revisit.

diegoboston commented on mygame.js:
```diff
@@ -25,4 +25,111 @@ Scene.prototype.bug = function scene_bug(message) {
\ No newline at end of file
+}
+
+/*  AUTOMATIC CHOICE SELECTION #1
```
I am not sure I understand why you have #1 #2 #3, it seems like you have the code here (step #1) and the code needed in startup.txt (#2).

diegoboston commented on mygame.js:
```diff
@@ -25,4 +25,111 @@ Scene.prototype.bug = function scene_bug(message) {
+    separated by a comma, if your choices are nested.
+
+    A few helper variables are also created to help confirm that the length/number
+    (cslib_choice_count) of options and their text content (cslib_choice_n) is as expected.
```
didn't you just say this above?

CareyJWilliams replied:
I can try to consolidate this a little more.

diegoboston commented on mygame.js:
```diff
@@ -25,4 +25,111 @@ Scene.prototype.bug = function scene_bug(message) {
+    the minimized code, please remember to update the source below as well.
+*/
+/*
+    Scene.prototype.choice = function choice(data) {
```
so, this must be copied to startup.txt? 

as I mentioned in the past, an alternative is to have EVERYTHING in startup.txt (even if hard to read, easier if we have the preprocessor)

CareyJWilliams replied:
Hard to read (for users wanting to learn, not just contributors), near impossible to track (useful) changes through version control? Nope, sorry, I'm not sold. I'd sooner suggest forking ChoiceScript and making subtle changes, honestly.

diegoboston replied:
Agreed

diegoboston commented on scenes/startup.txt:
```diff
@@ -63,6 +84,10 @@
 
+*if choice_randomtest
+	*comment override randomtest's Scene.prototype.choice so we can perform automated testing (original source in mygame.js)
+	*script Scene.prototype.choice=function(i){var t=["choice"];i&&(t=i.split(/ /));var n=this.parseOptions(this.indent,t),e=this;timeout=function(){println(""),e.renderOptions(t,n,function(i){e.standardResolution(i)})},this.finished=!0};
```
that's where you need the preprocessor :-)

danspinola reviewed (dismissed):
As I far as I could tell, it looks awesome. 😁 But I'm not too familiar with JS, so I might have missed something.

diegoboston commented on mygame.js:
```diff
@@ -25,4 +25,111 @@ Scene.prototype.bug = function scene_bug(message) {
+      throw new Error("Error: stuck in a choice menu but selections (cslib_choice_select) were exhausted.");
+
+    // Extract the next choice selection
+    var selections = this.stats.cslib_choice_select.split(",");
```
do you need to support nested choices in your tests?

CareyJWilliams replied:
Nested is probably the wrong word... 'subsequent' (though that would include nested).
But yes, this way I can check pagination works (both in terms of navigation, and that pages past page 1 are correct).
It's actually really quite fun...

diegoboston commented on scenes/cslib_menu.txt:
```diff
@@ -105,38 +107,54 @@
+*temp prefix "param"
+*temp filter ""
+*temp filter_value ""
+*temp page_count "@{(param_count <= const_items_per_page) 1|${(param_count / const_items_per_page)}}"
```
This stuff is getting really fancy - and it's really tricky for me to follow. Let's just test this stuff in and out.

CareyJWilliams replied:
It's basically a ternary operator: page_count = (param_count <= page_length) ? 1 : (param_count / page_length).
Happy to change variable names if that would help.

diegoboston commented on scenes/cslib_menu.txt:
```diff
@@ -105,38 +107,54 @@
-*comment swap the next/prev buttons intelligently for smoother navigation
-*temp next_preference true
+*label _init_state
+*comment determines the next/prev option position for smoother navigation
```
Again, this is very ambitious. A version with no pagination would be easier to understand and debug. BTW, do you really expect a menu with multiple pages of choices?

CareyJWilliams replied:
See my other comment somewhere above: I basically think that removing pagination would make this module redundant. Maybe you need to actually play with this patch yourself? Because I actually found it surprisingly easy to test pagination. It's quite fun actually.

You don't need to use the test framework, you can just fire up a normal choice, works a charm.

```
    *comment configure automatic choice selection
    *set cslib_catch_choice true
    *comment dictate which options (comma separated, for nested choices) should be selected
    *set cslib_choice_select "3"
    *fake_choice
      # Go east
        Eastward bound.
      # Go west
        Westward bound.
      # Go south
        Southward bound.
    *comment 'Southward bound.' will be printed here.
    *comment And helper variables will be populated:
    ${cslib_choice_count} == 3
    ${cslib_choice_1} == "Go east."
    ${cslib_choice_2} == "Go west."
    *comment etc.
    *comment and turn off automatic choice selection
    *set cslib_catch_choice false
```

diegoboston replied:
thanks!

diegoboston commented on scenes/cslib_menu.txt:
```diff
@@ -149,121 +167,202 @@
-	*bug cslib_menu: menu construction failed as there were 0 matching items — consider allowing cancellation
+*temp total_item_count 0
+*temp cur_option_index 1
+*temp cur_item_index "@{const_invert_order ${{prefix&\"_count\"}}|1}"
```
Is this a multi-replace in a string? Wow, we are pushing the envelope here. This is actually a very powerful operator.

CareyJWilliams replied:
Yeah, basically a ternary (again).

diegoboston commented on scenes/cslib_menu.txt:
```diff
@@ -149,121 +167,202 @@
+*label filter_loop
+*if (filter != "")
+	*gosub _apply_filter {((prefix&"_")&(filter&"_"))&cur_item_index} filter_value
+	*if not(cslib_ret)
```
does this mean cslib_ret != ""  ? I really lost track here :-(

CareyJWilliams replied:
_apply_filter returns a boolean, so it's guarding on the result of that being false.

diegoboston commented on scenes/cslib_menu.txt:
```diff
@@ -149,121 +167,202 @@
+*gosub _init_option_vars
+*label _populate_page_loop
+*if (slot <= const_items_per_page)
+	*if (prefix = "param")
```
Make sure you have tests both for arrays and params. Alternatively, we can choose to launch with a more limited (and better tested) set of functionality.

CareyJWilliams replied:
Test framework is here now, so feel free to suggest/write more tests. Should be very straight-forward now.

diegoboston commented on scenes/cslib_menu.txt:
```diff
@@ -149,121 +167,202 @@
+		*if (option_guard[slot])
+			*set page_item_count + 1
+			*set option_text[slot] "${{(prefix&\"_\")&{\"option_slot_id_\"&slot}}}"
+			*set option_ret_val[slot] option_slot_id[slot]
```
this is REALLY hard to follow, CS doesn't help

CareyJWilliams replied:
I'm open to suggestions...

diegoboston commented on scenes/cslib_menu.txt:
```diff
@@ -149,121 +167,202 @@
 *return
 
+
+*label _apply_filter
```
did you decide to keep the filters?

CareyJWilliams replied:
Yes.

diegoboston commented on scenes/test_menu.txt:
```diff
@@ -0,0 +1,173 @@
+*set cslib_choice_select "1"
+*set cslib_catch_bug true
+*gosub_scene cslib_menu build_simple 1 2 3 4 5 6
+*set cslib_catch_bug false
```
do you want to assert on the message?

CareyJWilliams replied:
Oops, yes. Good catch :)

diegoboston replied:
I can't see the updated version with the fixed test

diegoboston commented on scenes/test_menu.txt:
```diff
@@ -0,0 +1,173 @@
+
+*gosub test_start "simple - multi-page (navigation)"
+*comment Select 'Next' then 'Next' then 'Previous', then '8' on Page 2
+*set cslib_choice_select "2,2,2,5"
```
oooh, you need multiple choices for pagination, right?

CareyJWilliams replied:
Correct.

diegoboston commented on scenes/test_menu.txt:
```diff
@@ -0,0 +1,173 @@
+
+*comment :::::: ARRAY tests ::::::
+*comment array - init test data
+*set test_menu_opt_count 7
```
I won't stop this, but the filters are adding a layer of complexity - we don't have a strong design for 'objects' yet (for instance, I would feel this more natural: menu_opt[2]["visible"] than menu_opt["visible"][2])

CareyJWilliams replied:
If we had an array module that allowed you to filter values _before_ passing them to cslib_menu, then I might agree, but as it stands there is no other way to present only a subset of options -- which is a crucial feature.

We even use it in library testing framework itself: https://github.com/ChoicescriptIDE/cslib/blob/9e69e00e65399c239ec5f9cb023fb7abc9b0b1ee/scenes/_test_entry.txt#L61 -- To only display test logs with content.

Removing this would make cslib_menu useless outside of situations where you knew every single option was selectable (which I imagine is actually the less frequent of use-cases).

--

The whole thing is meant to be a "beta" right? So let's see what people think? Better for people to say they don't like it and remove it, then never get feedback on it at all.

diegoboston replied:
Agreed

diegoboston commented on scenes/test_menu.txt:
```diff
@@ -0,0 +1,173 @@
+
+*comment *gosub test_start "array - filter"
+*comment *set cslib_choice_select "2"
+*comment *gosub_scene cslib_menu build_array "test_menu_opt" false *comment "visible" "MATCH" true
```
are you missing a newline?

CareyJWilliams replied:
Huh, not sure what I've done there. Will take a look, thank you.

diegoboston commented on scenes/test_menu.txt:
```diff
@@ -0,0 +1,173 @@
+*comment *gosub_scene cslib_menu build_array "test_menu_opt" false *comment "visible" "MATCH" true
+*comment *gosub test_assert_equal 3 cslib_choice_count
+*comment *gosub test_assert_equal 1 cslib_ret
+*comment *gosub test_assert_equal "Even" cslib_choice_1
```
are you not testing filters?

CareyJWilliams replied:
Apparently not. That's embarrassing, but probably what you get for pushing code at stupid o'clock.
Thanks for catching, I shall fix.

diegoboston reviewed (commented):
This is very advanced, congratulations. Very nice trick to test the menu.
On the other side, this code is VERY hard to follow due to CS limitations --- great to have extensive tests.

I would consider launching with something simpler (without pagination or filters) but I won't stop you.   :-)

CareyJWilliams committed:
Add cslib_array module

This module provides basic utilities
for manipulating "cslib arrays", which
are cslib "arrays" with _max and _count
attributes.

Initial functionality includes add/remove,
clone, filter and set/clean routines.

CareyJWilliams committed:
Remove cslib's filter functionality

Removing cslib's filter functionality
greatly reduces the complexity of the
module, for both contributors and users.

The expected usage is to now filter
an array with cslib_array before passing
it to cslib_menu.

diegoboston commented on scenes/cslib_array.txt:
```diff
@@ -0,0 +1,215 @@
+*bug cslib_array is a cslib module and must not be accessed directly
+
+*comment TODO: Decide array_attribute structure
```
remove if not used

diegoboston commented on scenes/cslib_array.txt:
```diff
@@ -0,0 +1,215 @@
+    *if (src_index > tar_max)
+        *bug cslib_array.filter_by_routine: Target array '${p_array_ref_tar}' of max size ${tar_max} is too small to hold all filtered values
+    *comment pass current element to filter function
+    *if (filter_by_routine)
```
if we are not (yet) doing filter by value, I would simplify the code here - YAGNI   :-)

https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it

diegoboston commented on scenes/cslib_array.txt:
```diff
@@ -0,0 +1,215 @@
+*temp n 1
+*temp src_count {p_array_ref_src&"_count"}
+*temp tar_max {p_array_ref_tar&"_max"}
+*if (src_count > tar_max)
```
great!

diegoboston commented on scenes/cslib_array.txt:
```diff
@@ -0,0 +1,215 @@
+*comment Sets all elements in a given array to a given value,
+*comment and sets its _count to _max.
+*comment ------------------------------
+*label set_all
```
this is very handy, I like it!

diegoboston commented on scenes/cslib_array.txt:
```diff
@@ -0,0 +1,215 @@
+*comment ------------------------------
+*label set_all
+*params p_array_ref p_value
+*temp end "@{param_count=3 ${param_3}|${{p_array_ref&\"_max\"}}}"
```
VERY NICE!

diegoboston commented on scenes/cslib_array.txt:
```diff
@@ -0,0 +1,215 @@
+*gosub insert p_array_ref p_value ({{p_array_ref&"_count"}} + 1)
+*return
+
+*comment INSERT
```
I was going to say: next we need some push/pop operation!

diegoboston commented on scenes/cslib_array.txt:
```diff
@@ -0,0 +1,215 @@
+*return
+
+
+*comment SHIFT
```
maybe add some more documentation on these operations

diegoboston commented on scenes/cslib_menu.txt:
```diff
@@ -105,165 +77,173 @@
+*if (slot <= const_items_per_page)
+	*comment special handling for simple menus
+	*temp offset slot+(const_items_per_page * (cur_page -1))
+	*set option_guard[slot] ({prefix&"_count"} >= offset) and (const_items_per_page >= slot)
```
non actionable comment :-)

It's unfortunate we can't do {array_name}[idx], and that we have to bother with the underscore _ every time. It makes it unnecessarily hard to read

CareyJWilliams replied:
I've moved this out into an 'arr_count' temp, so it's a bit better now.

diegoboston commented on scenes/cslib_menu.txt:
```diff
@@ -105,165 +77,173 @@
+	*set option_guard[slot] ({prefix&"_count"} >= offset) and (const_items_per_page >= slot)
+	*if (option_guard[slot])
+		*set page_item_count + 1
+		*set option_text[slot] "${{prefix&\"_${offset}\"}}"
```
would this work?
*set option_text[slot] {prefix & "_${offset}" }

also, if you change the prefix to include the _ (new *temp prefix_)  you can simply do

*set option_text[slot] {prefix_ & offset}

CareyJWilliams replied:
Thanks for this, it's a really good suggestion.
So I've added the _ to the prefix variable, and also moved the array count out into a variable. It's referenced a lot, and was indeed very ugly to look at.

Things look much better now:

```
*temp arr_count {{prefix&"count"}}
*temp page_count "@{(arr_count <= const_items_per_page) 1|${(arr_count / const_items_per_page)}}"
```

diegoboston commented on scenes/test_array.txt:
```diff
@@ -0,0 +1,243 @@
+*gosub test_assert_equal "JACKPOT" test_arr_7
+*gosub test_assert_equal "JACKPOT" test_arr_8
+*gosub test_assert_equal "JACKPOT" test_arr_9
+*gosub test_assert (((test_arr_max = test_arr_count)) and (test_arr_max = 9)) "test_arr_max (${test_arr_max}) and test_arr_count (${test_arr_count}) should both be equal to 9"
```
consider using multiple assert (instead of AND), so you know which one failed

CareyJWilliams replied:
Because it uses a custom assert message this actually makes it more obvious which one failed (the values are printed next to each), versus two generic "expected 9, received ?".

diegoboston commented on scenes/test_array.txt:
```diff
@@ -0,0 +1,243 @@
+
+*comment our comparison routine applied to each array element to *comment determine whether it should be included in the filtered result
+*goto _skip_filter_test_is_even
+*label _filter_test_is_even
```
NICE!

diegoboston replied:
NICE!

diegoboston commented on scenes/test_array.txt:
```diff
@@ -0,0 +1,243 @@
+*gosub_scene cslib_array filter_by_routine "const_num" "test_arr" "test_array" "_filter_test_is_even"
+*set cslib_catch_bug false
+*temp test_arr_max test_arr_max_store
+*gosub test_finish "Bug: cslib_array.filter_by_routine: Target array 'test_arr' of max size 3 is too small to hold all filtered values" cslib_bug_message 
```
Very thorough tests, well done!

diegoboston reviewed (dismissed):
This is really great. Please see what comments are actionable before merging.

CareyJWilliams committed:
Test examples

Run randomtest on the examples in the example
folder, to increase confidence and ensure
they stay up-to-date with the APIs.

diegoboston commented on .github/workflows/test-examples.yml:
```diff
@@ -0,0 +1,31 @@
+      - name: Test Examples
+        run: |
+          cp cslib/examples/test-examples.sh .
+          ./test-examples.sh
```
this is really neat!

diegoboston commented on scenes/cslib_menu.txt:
```diff
@@ -105,165 +77,174 @@
+	*bug cslib_menu._init_state: const_items_per_page cannot be greater than 10
+*comment the text displayed on the cancel/return button
+*temp const_cancel_text "[return]"
+*temp arr_count {{prefix&"count"}}
```
do you need double braces?

CareyJWilliams replied:
No... Good spot. Although, if it's alright, I'll tidy that up in a separate commit. It's still functionally correct.
